index.js

```diff
@@ -0,0 +1,414 @@
+  return amount;
+};
+
+/**
+ * Gets a PayPal auth token so we can call on the api.
+ *
+ * @param {object} options - Options to use.
+ * @param {string} [options.clientId = paypal_client_id] - PayPal Client Id.
+ * @param {string} [options.secret = paypal_secret] - PayPal Secret.
+ *
+ * @returns {Promise<object>} The data returned with an `access_token`.
+ */
+const getAuthToken = async ({clientId, secret, api}) => {
+  const authUrl = `${api}/v1/oauth2/token`;
+  const options = {
+    headers: {
+      Accept: 'application/json',
+      'Content-Type': 'application/x-www-form-urlencoded',
+      'Accept-Language': 'en_US'
+    },
+    auth: {
+      username: clientId,
+      password: secret
+    }
+  };
+  const body = 'grant_type=client_credentials';
+  const {data} = await axios.post(authUrl, body, options);
+  return data;
+};
+
+/**
+ * Gets the options needed to make a secure paypal request.
+ *
+ * @param {object} options - Options to use.
+ * @param {string} [options.clientId = paypal_client_id] - PayPal Client Id.
+ * @param {string} [options.secret = paypal_secret] - PayPal Secret.
+ *
+ * @returns {Promise<object>} Settings for paypal.
+ */
+const getOptions = async ({clientId, secret, api} = getConfig()) => {
+  const {access_token, token_type} = await getAuthToken(
+    {clientId, secret, api});
+  const Authorization = `${token_type} ${access_token}`;
+  const headers = {
+    Authorization,
+    Accept: 'application/json'
+  };
+  return {api, headers};
+};
+
+/**
+ * Gets an order from PayPal to ensure the order was made.
+ *
+ * @param {object} options - Options to use.
+ * @param {object} options.order - The order from the client.
+ *
+ * @returns {Promise<object|null>} The order data from PayPal.
+ */
+const getOrder = async ({id}) => {
+  try {
+    const {headers, api} = await getOptions();
+    const options = {headers};
+    const url = `${api}/v2/checkout/orders/${id}`;
+    const {data} = await axios.get(url, options);
+    return data;
+  } catch(e) {
+    logger.error('getOrder not found', {error: e});
+    if(e.response.status === 404) {
+      return null;
+    }
+    throw e;
+  }
+};
+
+/**
+ * Takes in a payment object and uses the service id
+ * to find it in the payment gateway.
+ *
+ * @param {object} options - Options to use.
+ * @param {object} options.payment - A payment object.
+ *
+ * @throws {BedrockError} - Throws not NotFound.
+ *
+ * @returns {Promise<object>} The order from the payment gateway.
+ */
+const getOrderFromPayment = async ({payment}) => {
+  const order = await getOrder({id: payment.serviceId});
+  if(!order) {
+    const message = 'PayPal order not found.';
+    payment.status = PaymentStatus.FAILED;
+    payment.error = message;
+    await paymentService.db.save({payment});
+    throw new BedrockError(
+      message, Errors.NotFound, {httpStatusCode: 404, public: true});
+  }
+  return order;
+};
+
+/**
+ * Delete order only works on orders where no money has been exchanged.
```

Comment:
If the status is not one of the required types do they return something other than a 200 response?  Otherwise, should data be inspected here to determine success/failure?  We probably don't want clients/users interacting with raw paypal responses.

Reply:
Not sure I'm completely following this. You are saying why don't we check the order's status on PayPal before deleting an order?

Reply:
When you say "Delete order only works", what does that mean?  Does that mean that the paypal APIs are expected to produce a non 200 status code if those conditions are not met?  Are we handling those cases properly?

Reply:
you are correct. PayPal's API will throw if the order status is not `CREATED` or `AUTHORIZED`.
This is only called by processGateWayPayment if the PayPal order's status is `CREATED`. 
It could throw if somehow the status or the order were to change before the delete action happens. In that case deleteOrder will throw then processGateWayPayment will throw, but the payment can be resumed later. If on resume the PayPal order is now set to COMPLETED the order will process as normally, other wise it will be marked as FAILED. It's not possible to change the PayPal order status after it leaves the CREATED or AUTHORIZED statuses.

Reply:
Forgive me if it feels like I am repeating things already covered. I am still unclear.

> "Delete order only works on orders where no money has been exchanged."

1. Who controls whether or not "delete order" works? Paypal's code or our code?
2. It doesn't make sense when you say PayPal's API throws an error since it's an HTTP Request. What HTTP Error Codes are thrown when the conditions are not met?
3. We have a single catch-all for any type of error and then we throw. Should all errors be treated the same?

Reply:
@gannan08 

1. papypal controls which delete operations fail:

> Cancels an order, by ID, and deletes the order. To call this method, the order status must be CREATED or APPROVED.

2. I would consider HTTP status codes from an API endpoint that are above the 400 level to be an error.

https://developer.paypal.com/docs/api/orders/v1/#orders-cancel-response
> A successful request returns the HTTP 204 No Content status code with no JSON response body. If the order is already paid, the order cannot be canceled and the request returns the HTTP 422 Unprocessable Entity status code with the message, This order is in progress.

3. We have to treat axios errors differently because they contain the Authorization header with an oath token that lasts an hour. Hence it would be possible to use our paypal oauth token for an hour if one were to get it from the logs. So not so sure on a one size fits all solution here.